Gradient Descent/gradient.cpp

```diff
@@ -0,0 +1,156 @@
+#include <math.h> 
+#include <numeric>
+#include <cstdlib>
+
+using namespace std;
+
+const double epsilon = pow(10,-6); // Fix a small epsilon for finite difference scheme to calculate grad
+
+// This is the function to be minimised. At the moment it is f(x) = x^2
+double objfun(vector<double>& x){               // Takes a vector x...
+    double output_val = 0.0;
+    for(int i=0 ; i < x.size(); i++){
+        output_val += pow(x[i],2);}             //... and computes x_i^2 for each coordinate...
+    return output_val;                      //... and returns vector of calculation.
+}
+
+// This another function to be minimised. Uncomment out if
+// you want to use it. At the moment it is f(x) = x^2 + x^4 + 15
+//double objfun(vector<double>& x){               // Takes a vector x...
+//    double output_val = 0.0;
+//    for(int i=0 ; i < x.size(); i++){
+//        output_val += pow(x[i],2) + pow(x[i],4);}             //... and computes x_i^2 +x_i^4 + 15 for each coordinate...
+//    return output_val + 15;                      //... and returns vector of calculation.
+//}
+
+
+// Function to calculate the gradient of the objective function
+vector<double> grad(vector<double>& input_vals){        // Takes a vector of type double as input values
+    vector<double> grad_out;                            // Gradient vector to be returned
+    vector<double> x_up = input_vals;                   // Vector to keep x_i + epsilon in the ith component
+    vector<double> x_down = input_vals;                 // Vector to stall x_i - epsilon in the ith component
+    for(int i=0; i < input_vals.size(); i++){        // Loop calculates derivative in each direction...
+        x_up[i] += epsilon;                          //...using a finite difference scheme...
+        x_down[i] -= epsilon;
+        grad_out.push_back((objfun(x_up) - objfun(x_down)) / (2 * epsilon));  // (this epsilon was fixed above)
+        x_up[i] -= epsilon;
+        x_down[i] += epsilon;
+    }
+    return grad_out;                                 //... and then returns the gradient vector
+}
+
+// Function to print vectors of any length
+void print_vector(vector<double> data)
+{
+    for(int i = 0 ; i < data.size(); i++)
+    {
+        cout << data[i] << "\t";
+    }
+    cout << endl;
+}
+
+// Function to find the L2 norm of a vector
+double l2_norm(vector<double> &x)
+{
+    double norm = 0.0;
+    for(int i=0; i<x.size();i++){
+        norm += x[i] * x[i];
+    }
+    return sqrt(norm);
+}
+
+// Function to subtract vector y from vector x
+vector<double> subtract_vect(vector<double> x, vector<double> y){
+    vector<double> output;
+    for(int i = 0; i < x.size();i++){
+        output.push_back(x[i]-y[i]);
+    }
+    return output;
+
+}
+
+// Function to multiply every element in a vector x by a scalar k
+vector<double> multiplication_vectorscal(vector<double> &x, double k){
+    for (int i = 0; i < x.size() ; i++)
+    {
+        x[i] *= k;
+    }
+    return x;
+}
```

Comment:
All these helper functions can be put in a separate header file -- I think most will be handled by Eigen